iree/runtime/demo/hello_world_explained.c

```diff
@@ -0,0 +1,273 @@
+    status = iree_runtime_demo_perform_mul(session);
+  }
+
+  // Release the session and free all resources.
+  iree_runtime_session_release(session);
+  return status;
+}
+
+//===----------------------------------------------------------------------===//
+// 3. Call a function within a module with buffer views
+//===----------------------------------------------------------------------===//
+// The inputs and outputs of a call are reusable across calls (and possibly
+// across sessions depending on device compatibility) and can be setup by the
+// application as needed. For example, an application could perform
+// multi-threaded buffer view creation and then issue the call from a single
+// thread when all inputs are ready. This simple demo just allocates them
+// per-call and throws them away.
+
+// Sets up and calls the simple_mul function and dumps the results:
+// func @simple_mul(%arg0: tensor<4xf32>, %arg1: tensor<4xf32>) -> tensor<4xf32>
+//
+// NOTE: this is a demo and as such this performs no memoization; a real
+// application could reuse a lot of these structures and cache lookups of
+// iree_vm_function_t to reduce the amount of per-call overhead.
+static iree_status_t iree_runtime_demo_perform_mul(
+    iree_runtime_session_t* session) {
+  // Initialize the call to the function.
+  iree_runtime_call_t call;
+  IREE_RETURN_IF_ERROR(iree_runtime_call_initialize_by_name(
+      session, iree_make_cstring_view("module.simple_mul"), &call));
+
+  // Append the function inputs with the HAL device allocator in use by the
+  // session. The buffers will be usable within the session and _may_ be usable
+  // in other sessions depending on whether they share a compatible device.
+  iree_hal_allocator_t* device_allocator =
+      iree_runtime_session_device_allocator(session);
+  iree_status_t status = iree_ok_status();
+  {
+    // %arg0: tensor<4xf32>
+    iree_hal_buffer_view_t* arg0 = NULL;
+    if (iree_status_is_ok(status)) {
+      static const iree_hal_dim_t arg0_shape[1] = {4};
+      static const float arg0_data[4] = {1.0f, 1.1f, 1.2f, 1.3f};
+      status = iree_hal_buffer_view_wrap_or_clone_heap_buffer(
+          device_allocator,
+          // Shape dimensions and rank:
+          arg0_shape, IREE_ARRAYSIZE(arg0_shape),
+          // Element type:
+          IREE_HAL_ELEMENT_TYPE_FLOAT_32,
+          // Where to allocate (host or device):
+          IREE_HAL_MEMORY_TYPE_HOST_LOCAL | IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE,
+          // What access to allow to this memory (this is .rodata so READ only):
+          IREE_HAL_MEMORY_ACCESS_READ,
+          // Intended usage of the buffer (transfers, dispatches, etc):
+          IREE_HAL_BUFFER_USAGE_ALL,
+          // The actual heap buffer to wrap or clone and its allocator:
+          iree_make_byte_span((void*)arg0_data, sizeof(arg0_data)),
+          iree_allocator_null(),
+          // Buffer view + storage are returned and owned by the caller:
+          &arg0);
```

Comment:
This is still pretty verbose for a higher level API. The comments help, but I'm still wondering in what cases / how frequently users will want to actually change the memory type, memory access, and buffer usage. Could we add a helper that uses these as defaults?

Reply:
Yeah it's a lot of characters, but easy to copy and intellisense works really well. This isn't a higher-level API - it's the HAL API. And if there's one place to have verbosity when using IREE it's for buffer management - buffers are kind of important :)

There are no defaults that are good for this kind of stuff - you _must_ know whether you want your memory to live on the host or the device and you _must_ know if you want it read-only or read-write. What's cheap on the local CPU device or GPU with unified memory can be catastrophically slow on a sandbox/enclave or functionally impossible on a GPU with discrete memory (see our out-of-memories with HOST_LOCAL|DEVICE_VISIBLE memory on AMD after only 256MB of allocations). The alternative is to make multiple functions with names like `iree_hal_buffer_view_wrap_or_clone_heap_buffer_on_device_and_readonly`, which is just API bloat and makes changing things in the future harder. This way, new enums can be added without needing to plumb through new API calls.

Buffer usage could _maybe_ be defaulted to ALL, however once you have the other two I don't see a reason to skip it - and IREE_HAL_BUFFER_USAGE_CONSTANT can be extremely useful (indicates that the memory will never change out from under the HAL so when mirrored on to a device/remote host/sandbox it's safe to assume it'll never change, etc). If we add more flags here (like EXPORTABLE or ENCYRPTED or whatever) we don't want to have to plumb through new methods for taht.

At the point you want something higher level than this you move out of the realm of things we care about in core and into binding land and I'd rather have this API be something bindings could use with full fidelity. Trying to hand-hold here just leads to cases where people have to choose between using the nicer API _or_ having access to that one lower-level flag that lets them do what they need. I'm trying not to force that tradeoff :)

Basically: if you are using C you can deal with 3 extra arguments that are load-bearing to correctness and performance. If you don't care, you can use something like the tflite API - and that's a pretty good example of the mess I'm trying to avoid here (see EnsureTensorDataIsReadable, GetBufferHandle, SetAllowBufferHandleOutput, SetCustomAllocationForTensor, etc).

Reply:
(I'm not sure the best way to say all that in the docs/comments, tried not to expound too much in here as I can tend to type too much and I didn't want it to be confusing, but I'm personally a fan of this kind of guidance so open to suggestions on how to best integrate it :)